### PR TITLE
Fix upload

### PR DESCRIPTION
**Pull request by a-levra**
Screenshot 2023-10-29 at 01 01 16

### CODE REVIEW

**Comment by a-levra** [comment]
ATTENTION : I'm currently adding features on this branch. Do not delete this branch.

File: src/main.cpp [hunk 1]
```diff
@@ -14,6 +16,16 @@ static void handleSignal(int signum);
...
+	int pid = fork();
+	if (pid < 0)
+		return EXIT_FAILURE;
+	if (pid == 0){
+		char *file = strdup("/app/updateHTML.sh");
+		char *args[] = {file, NULL};
+		execve(file, args, NULL);
+		return EXIT_FAILURE;
+	}
+
```
Comment (tdameros) [inline src/main.cpp]: It's a good debugging feature, but I'm not sure if it's necessary to hard code tests like this on the main branch. In the future, these kinds of features will be directly supported by the CGI.

Also, be careful, this creates an infinite while loop that consumes resources even when the page is not requested (which won't be the case with the CGI).

Lastly, from a technical standpoint, strdup is not secure; it's preferable to use a C++ string with c_str(). The child process is also never waited for or killed, which leads to process leaks (very dangerous).

That's why I think it's better to keep this feature on a debug branch and not on the main branch.

**Review by tdameros — CHANGES_REQUESTED** [review]

File: src/main.cpp (hunk 1 above)
Comment (a-levra) [reply src/main.cpp]: This is not a debugging feature
In the latest version, the script is only call when needed (no loop)
strdup seems to be required for this particular usage, but i'm searching if i can do it in other way.

Comment (tdameros) [reply src/main.cpp]: OK, but why "hard code" the launch of a bash script from our C++ code when CGIs can handle all .sh files directly?

Comment (a-levra) [reply src/main.cpp]: OK, i understand. Let keep this here, and update when CGIs will be implemented !

Comment (a-levra) [reply src/main.cpp]: i'm not finished for tonight, i'll request another review later

Comment (tdameros) [reply src/main.cpp]: Okay, no problem !

File: src/utils/createFile.cpp [hunk 2]
```diff
@@ -5,8 +5,10 @@
 #include "logger/logging.hpp"
 
 void createFile(const std::string &filename, const std::string &fileContent) {
+
 	coloredLog("Creating file: ", filename, GREEN);
```
Comment (tdameros) [inline src/utils/createFile.cpp]: Replace coloredLog by logging

File: src/utils/createFile.cpp [hunk 3]
```diff
@@ -5,8 +5,10 @@
...
 //	coloredLog("File content: ", fileContent, GREEN);
+
 	std::ofstream file;
 	file.open(("uploadedFiles/" + filename).c_str());
 	file << fileContent;
```
Comment (tdameros) [inline src/utils/createFile.cpp]: secure open ?

using this code :

std::ofstream file("filename");
if (file.is_open())....

File: src/HttpMessages/HttpResponse.cpp [hunk 4]
```diff
@@ -118,9 +124,23 @@ void HttpResponse::generateBody(Location &location) {
...
 	getFileFromPostAndSaveIt();
+	updateHTML();
 	this->buildGet(location);
 }
 
+void HttpResponse::buildDelete(Location &location) {
+	coloredLog("Building DELETE response: ", "", BLUE);
+	_request.displayRequest();
+	std::string res = getResource(location);
+	coloredLog("resources : " ,res, BLUE);
```
Comment (tdameros) [inline src/HttpMessages/HttpResponse.cpp]: replace coloredLog

File: src/HttpMessages/HttpResponse.cpp [hunk 5]
```diff
@@ -118,9 +124,23 @@ void HttpResponse::generateBody(Location &location) {
 
 void HttpResponse::buildPost(Location &location) {
 	getFileFromPostAndSaveIt();
+	updateHTML();
 	this->buildGet(location);
 }
 
+void HttpResponse::buildDelete(Location &location) {
+	coloredLog("Building DELETE response: ", "", BLUE);
```
Comment (tdameros) [inline src/HttpMessages/HttpResponse.cpp]: replace coloredLog by logging

File: src/HttpMessages/HttpResponse.cpp [hunk 6]
```diff
@@ -189,9 +209,10 @@ void HttpResponse::GenerateErrorBody() {
 	coloredLog("Error page generated: ", "", RED);
 	coloredLog("Error code: ", toString(_statusCode), RED);
 	coloredLog("Error message: ", _statusMessage, RED);
```
Comment (tdameros) [inline src/HttpMessages/HttpResponse.cpp]: same

File: src/HttpMessages/HttpResponse.cpp [hunk 7]
```diff
@@ -200,7 +221,7 @@ void HttpResponse::GenerateErrorBody() {
 }
 
 void HttpResponse::getFileFromPostAndSaveIt() {
-	coloredLog("Building POST response: ", "", BLUE);
+	coloredLog("Building POST response ", "", BLUE);
```
Comment (tdameros) [inline src/HttpMessages/HttpResponse.cpp]: same

File: src/utils/createFile.cpp [hunk 8]
```diff
@@ -5,10 +5,12 @@
...
-	logging::debug("Creating file: " + filename);
-	logging::debug("File content: " + fileContent);
+
+	coloredLog("Creating file: ", filename, GREEN);
+//	coloredLog("File content: ", fileContent, GREEN);
+
 	std::ofstream file;
-	file.open(("uploaded_files/" + filename).c_str());
+	file.open(("uploadedFiles/" + filename).c_str());
 	file << fileContent;
```
Comment (tdameros) [inline src/utils/createFile.cpp]: Secure open

**Review by tdameros — CHANGES_REQUESTED** [review]

File: src/utils/updateHTML.cpp [hunk 9]
```diff
@@ -0,0 +1,26 @@
...
+	int pid = fork();
+	if (pid < 0)
+		return EXIT_FAILURE;
+	if (pid == 0){
+		char *file = strdup("/app/updateHTML.sh");
+		char *args[] = {file, NULL};
+		execve(file, args, NULL);
+		free(file);
+		return EXIT_FAILURE;
+	}
```
Comment (tdameros) [inline src/utils/updateHTML.cpp]: This fork never exits, so it launches several webservs

File: src/utils/updateHTML.cpp [hunk 10]
```diff
@@ -0,0 +1,26 @@
...
+#include <unistd.h>
+#include <stdlib.h>
+
+
+int updateHTML(){
+	coloredLog("updating HTML", "", RED);
+	coloredLog("updating HTML", "", BLUE);
+	coloredLog("updating HTML", "", YELLOW);
+	coloredLog("updating HTML", "", PURPLE);
+	coloredLog("updating HTML", "", GREEN);
```
Comment (tdameros) [inline src/utils/updateHTML.cpp]: Use logging module

**Review by tdameros — CHANGES_REQUESTED** [review]

**Review by tdameros — APPROVED** [review]

**Pull request merged**